### DISCUSSION TITLE
Table Export column order

### DISCUSSION THREAD

**Discussion by white3km** [yahoo/elide#3091] [Q&A]
Hello, 

Two questions/requests about Table Export column order.

1.  Normally Elide Json API responses list attributes in alphabetical order but with the table export the attributes seem to be in random order. I see that JsonExportFormatter is using Elide's ObjectMapper to write the response. Is there a way that the order can be defaulted to alphabetical like the normal responses?
2. We were considering a use case where the client could specify column order for table export.  Example:  /foo?fields[foo]=zeta,alpha,beta then the results return with columns in that order.  Would it be possible for you to update the TableExportFormatter interface to include the TableExportQuery parameter for the format method  (https://github.com/yahoo/elide/blob/master/elide-async/src/main/java/com/yahoo/elide/async/export/formatter/TableExportFormatter.java#L23)?

**Comment by aklish** [comment]
1. Should be doable with TableExportFormatter.
2. The preFormat method includes the EntityProjection which should have the fields in their requested order.

**Reply by white3km** [reply]
1. Right, I get that you can create your own Object Mapper for Alphabetical order (and maybe that's the solution). I was just trying to use Elide's Object Mapper since I figured it would already be configured to handle the entities in the same way as the rest of the api. That's why I was curious why Elide's Object Mapper doesn't use alphabetical order for the keys since normal elide responses return with attributes in alphabetical order.

2. Well EntityProjection has the requestedFields but since it's a Set it doesn't persist the order so I parse the query url from the TableExport parameter to get the requested field order. Both preFormat and postFormat methods have the TableExport so I could create a header and footer for the fields but I need the order in the format method. Trying to persist the order in the ExportFormatter feels brittle if we had multiple requests come in at the same time.

```
public class CSVExportFormatter implements TableExportFormatter {

	private CopyOnWriteArrayList<String> fieldList = new CopyOnWriteArrayList<>();

	public String preFormat(EntityProjection projection, TableExport query) {
		// Retrieve fields in query order
		List<String> fields = parseFieldsFrom(query.getQuery());
		
		if(fields.isEmpty()){
			fieldList.clear();
			// default field order
			return String.join(", ", projection.getRequestedFields()))
		} else {
			fieldList.addAll(fields);
			return String.join(", ", fields))
		}
	}
	
	public String format(PersistentResource resource, Integer recordNumber) {
                // Either uses fieldList or default order
		return generateCsv(resource, fieldList);
	}
}
```

**Comment by aklish** [comment]
EntityProjection fields are in an ordered Set (LInkedHashSet).  We could change the contract to make that explicit.

**Reply by white3km** [reply]
Hmm not seeing the same order as the query parameter though.

```
public String preFormat(EntityProjection projection, TableExport query) {
    log.info("query = " + query.getQuery());
    log.info("requested = " + String.join(", ", projection.getRequestedFields()));
}

Log:
query = /foo?page%5Btotals%5D=true&page%5Bnumber%5D=1&page%5Bsize%5D=500&fields%5Bfoo%5D=startDate,endDate,status,createdAt,createdBy,updatedAt,updatedBy
requested = updatedAt, startDate, updatedBy, endDate, createdBy, createdAt, status
```

**Comment by aklish** [comment]
Looks like @justin-tay Fixed this in #3099
